numpy/core/einsumfunc.py

```diff
@@ -0,0 +1,597 @@
+
+    Returns
+    -------
+    path : list
+        The greedy order of pair contractions.
+    """
+
+    path = []
+    for iteration in range(len(input_sets) - 1):
+        iteration_results = []
+        comb_iter = []
+
+        # Grab all unique pairs
+        for x in range(len(input_sets)):
+            for y in range(x + 1, len(input_sets)):
+                comb_iter.append((x, y))
+
+        for positions in comb_iter:
+
+            contract = _find_contraction(positions, input_sets, output_set)
+            idx_result, new_input_sets, idx_removed, idx_contract = contract
+
+            # Sieve the results based on memory_limit
+            if _compute_size_by_dict(idx_result, idx_dict) > memory_limit:
+                continue
+
+            # Build sort tuple
+            removed_size = _compute_size_by_dict(idx_removed, idx_dict)
+            cost = _compute_size_by_dict(idx_contract, idx_dict)
+            sort = (-removed_size, cost)
+
+            # Add contraction to possible choices
+            iteration_results.append([sort, positions, new_input_sets])
+
+        # If we did not find a new contraction contract remaining
+        if len(iteration_results) == 0:
+            path.append(tuple(range(len(input_sets))))
+            break
+
+        # Sort based on first index
+        iteration_results.sort()
+        best = iteration_results[0]
+        path.append(best[1])
+        input_sets = best[2]
+
+    return path
+
+
+# Rewrite einsum to handle different cases
+def einsum(*operands, **kwargs):
+    """
+    Evaluates the Einstein summation convention based on the operands,
+    differs from einsum by utilizing intermediate arrays to
+    reduce overall computational time.
+
+    Produces results identical to that of the einsum function; however,
+    the contract function expands on the einsum function by building
+    intermediate arrays to reduce the computational scaling and utilizes
+    BLAS calls when possible.
+
+    Parameters
+    ----------
+    subscripts : str
+        Specifies the subscripts for summation.
+    *operands : list of array_like
+        These are the arrays for the operation.
+    tensordot : bool, optional (default: True)
+        If true use tensordot where possible.
+    path : bool or list, optional (default: ``greedy``)
+        Choose the type of path.
+
+        - if a list is given uses this as the path.
+        - 'greedy' means a N^3 algorithm that greedyally
+            chooses the best algorithm.
+        - 'optimal' means a N! algorithm that tries all possible ways of
+            contracting the listed tensors.
+
+    memory_limit : int, optional (default: largest input or output array size)
+        Maximum number of elements allowed in an intermediate array.
+    return_path : bool, optional (default: False)
+        If true retuns the path and a string representation of the path.
+
+
+    Returns
+    -------
+    if return_path:
```

Comment:
Maybe it would be better to create another dedicated function `einsum_path` for computing the optimal path? It's an awkward API to make a function that returns something totally different if it gets a keyword argument.